### PR TITLE
names: add contrib vocabulary

### PR DESCRIPTION
**Pull request by ppanero**
- requires https://github.com/inveniosoftware/invenio-records-resources/issues/289
- closes https://github.com/inveniosoftware/invenio-vocabularies/issues/84
- triggers https://github.com/inveniosoftware/invenio-vocabularies/issues/98

**Notes for the reviewers**
- The PID provider posses some complexities, but in order to unlock the loading/cli work and merge the vocabulary, work on the provider/pids will be done as a separate issue in https://github.com/inveniosoftware/invenio-vocabularies/issues/94

### CODE REVIEW

File: invenio_vocabularies/contrib/names/schema.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from invenio_records_resources.services.records.schema import BaseRecordSchema
+from marshmallow import Schema, ValidationError, fields, post_load, \
+    validates_schema
+from marshmallow_utils.fields import IdentifierSet, SanitizedUnicode
+from marshmallow_utils.schemas import IdentifierSchema
+
+from .config import names_schemes
+
+
+# FIXME: this is a dup from rdm-records, shall we add it in the aff vocab?
```
Comment (slint) [inline invenio_vocabularies/contrib/names/schema.py:23]: Indeed it's weird that this is not in this module (since it's pretty much a Service schema validation rule for creating affiliations). Maybe, shelve and we see it can be addressed separately.

File: invenio_vocabularies/contrib/names/schema.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    @post_load
+    def calculate_name(self, data, **kwargs):
+        """Update names for person.
+
+        Fill name from given_name and family_name if person.
+        """
+        family_name = data.get("family_name")
+        given_name = data.get("given_name")
+        if family_name and given_name:
```
Comment (slint) [inline invenio_vocabularies/contrib/names/schema.py:89]: Shouldn't this check if there is already `name`, and in that case do nothing?

File: tests/contrib/names/conftest.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    """Full name data."""
+    return {
+        "name": "Doe, John",
+        "given_name": "John",
+        "family_name": "Doe",
+        "identifiers": [
+            {
+                "identifier": "0000-0001-8135-3489",
+                "scheme": "orcid"
+            }
```
Comment (slint) [inline tests/contrib/names/conftest.py:78]: I would add a GND as well for completeness.

**Review by slint — COMMENTED** [review]
Minor things regarding search and schema, but otherwise LGTM :+1:

File: invenio_vocabularies/contrib/names/jsonschemas/names/name-v1.0.0.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      "description": "URI or classification code as identifier - globally unique among all names schemes.",
+      "$ref": "local://definitions-v1.0.0.json#/identifier"
+    },
+    "pid": {
+      "$ref": "local://definitions-v1.0.0.json#/internal-pid"
+    },
+    "scheme": {
+      "description": "Identifier of the name scheme.",
+      "$ref": "local://definitions-v1.0.0.json#/identifier"
+    },
```
[the remark this answers is not in the document]
Comment (ppanero) [reply invenio_vocabularies/contrib/names/jsonschemas/names/name-v1.0.0.json:21]: The scheme is the "sub-vocabulary" or "vocabulary type". e.g. for subjects we have the global vocabulary, but then there are different inputs.
- Vocabulary: Subjects
    - Scheme: MeSH
    - Scheme: FOS
    - ...
 
Currently for affiliations it would be `RoR` indeed, and for ours `ORCiD`. Unless we do not want to distinguish different sources. It might make deduplication more complicated, but it already is. Knowing where the vocabulary items com from is useful IMO.

File: invenio_vocabularies/contrib/names/jsonschemas/names/name-v1.0.0.json [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    "affiliations": {
+      "description": "Affiliations of the person.",
+      "type": "array",
+      "items": {
+        "type": "object",
+        "additionalProperties": false,
+        "properties": {
+          "id": {
+            "$ref": "local://definitions-v1.0.0.json#/identifier"
```
[the remark this answers is not in the document]
Comment (ppanero) [reply invenio_vocabularies/contrib/names/jsonschemas/names/name-v1.0.0.json:39]: Nope, because it is a vocabulary. So while the affiliation itself will have more fields, the "reference" in the name will be:
- `id` if it is an affiliation existing in the DB
- `name` if a custom one

See for example the affiliation in [`rdm-records`](https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/records/jsonschemas/records/definitions-v2.0.0.json#L2)

TL;DR: They can have multiple identifiers, yes (in their inner data) but they are referenced by one (the PK in the DB, sort of -  pidstore style)

File: tests/contrib/names/conftest.py (hunk 3 above)
Comment (ppanero) [reply tests/contrib/names/conftest.py:78]: added

File: invenio_vocabularies/contrib/names/schema.py (hunk 1 above)
Comment (ppanero) [reply invenio_vocabularies/contrib/names/schema.py:23]: done https://github.com/inveniosoftware/invenio-vocabularies/issues/98

File: invenio_vocabularies/contrib/names/config.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+class NamesSearchOptions(SearchOptions):
+    """Search options."""
+
+    suggest_parser_cls = SuggestQueryParser.factory(
+        fields=[
+            'name^100',
+            'family_name^100',
+            'given_name^100',
+            'identifiers.identifier^20',
```
[the remark this answers is not in the document]
Comment (ppanero) [reply invenio_vocabularies/contrib/names/config.py:35]: I agree, I add `affiliations.name`, the `id` it is basically the dumper field of the relation or the name if custom.The `id` is an identifier that might not be searched for (Do we expect people to look by RoR?, the suggesters on ES for affiliations are not meant/optimized for that). 

It is tricky to test and I feel like between `^10` and default (`^1`) there is not much difference. Like the `^100` definitely work, but more granular cases are hard to test.

File: invenio_vocabularies/contrib/names/jsonschemas/names/name-v1.0.0.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppanero) [reply invenio_vocabularies/contrib/names/jsonschemas/names/name-v1.0.0.json:21]: To be taken into account when importing actual data in the system. Don't want to subtype/schemed it with `orcid`.

File: invenio_vocabularies/contrib/names/schema.py (hunk 2 above)
Comment (ppanero) [reply invenio_vocabularies/contrib/names/schema.py:88]: IRL:
- do not overwrite `name` 
- validate: either `name` either `family_name` and `given_name`

**Comment by slint** [comment]
Included in #100
